Compiler/src/typeinfer.jl

```diff
@@ -451,38 +455,15 @@ function finishinfer!(me::InferenceState, interp::AbstractInterpreter)
 
     # finish populating inference results into the CodeInstance if possible, and maybe cache that globally for use elsewhere
     if isdefined(result, :ci)
-        result_type = result.result
-        result_type isa LimitedAccuracy && (result_type = result_type.typ)
-        @assert !(result_type === nothing)
-        if isa(result_type, Const)
-            rettype_const = result_type.val
-            const_flags = is_result_constabi_eligible(result) ? 0x3 : 0x2
-        elseif isa(result_type, PartialOpaque)
-            rettype_const = result_type
-            const_flags = 0x2
-        elseif isconstType(result_type)
-            rettype_const = result_type.parameters[1]
-            const_flags = 0x2
-        elseif isa(result_type, PartialStruct)
-            rettype_const = result_type.fields
-            const_flags = 0x2
-        elseif isa(result_type, InterConditional)
-            rettype_const = result_type
-            const_flags = 0x2
-        elseif isa(result_type, InterMustAlias)
-            rettype_const = result_type
-            const_flags = 0x2
-        else
-            rettype_const = nothing
-            const_flags = 0x0
-        end
-        relocatability = 0x0
+        (; rettype, exctype, rettype_const, const_flags) = ResultForCache(result)
         di = nothing
         edges = empty_edges # `edges` will be updated within `finish!`
         ci = result.ci
-        ccall(:jl_fill_codeinst, Cvoid, (Any, Any, Any, Any, Int32, UInt, UInt, UInt32, Any, Any, Any),
-            ci, widenconst(result_type), widenconst(result.exc_result), rettype_const, const_flags,
-            first(result.valid_worlds), last(result.valid_worlds),
+        min_world, max_world = first(result.valid_worlds), last(result.valid_worlds)
+        ccall(:jl_fill_codeinst, Cvoid, (
+            Any, Any, Any, Any, Int32, UInt, UInt,
+            UInt32, Any, Any, Any),
+            ci, rettype, exctype, rettype_const, const_flags, min_world, max_world,
             encode_effects(result.ipo_effects), result.analysis_results, di, edges)
         if is_cached(me)
             cached_result = cache_result!(me.interp, result, ci)
```

Comment:
I think the main question for review is whether this PR makes codegen unsound. Right here, we publish the ABI that will be used by this CodeInstance, for use by the optimizer in creating invoke edges. Any future changes to that info can result in codegen becoming unsound (as well as thread-unsafe).